cmd/controller-manager/main.go

```diff
@@ -189,62 +189,75 @@ func main() {
-		// for the server
-		sigChan := make(chan interface{}, 10)
-		//debouncedChan := make(chan interface{}, 10)
+func setupWatchQueue(stopCh <-chan struct{}, configWatcher *watcher.K8sConfigMapWatcher) workqueue.Interface {
+	// watch for reconciliation signals, and grab configmaps, then update the running configuration
+	// for the server
+	sigChan := make(chan interface{}, 10)
 
-		// debounce events from sigChan, so we dont hammer apiserver on reconciliation
-		eventsCh := utils.Coalescer(EventCoalesceWindow, sigChan, stopCh)
+	queue := workqueue.NewRateLimitingQueue(&workqueue.BucketRateLimiter{Limiter: rate.NewLimiter(rate.Limit(0.5), 1)})
 
-		go func() {
-			for {
-				setupLog.Info("Launching watcher for ConfigMaps")
-				if err := configWatcher.Watch(sigChan, stopCh); err != nil {
-					switch err {
-					case watcher.ErrWatchChannelClosed:
-						// known issue: https://github.com/kubernetes/client-go/issues/334
-						setupLog.Info("watcher channel has closed, restart watcher")
-					default:
-						setupLog.Error(err, "unable to watch new ConfigMaps")
-						os.Exit(1)
-					}
-				}
+	go func() {
+		for {
+			select {
+			case <-stopCh:
+				queue.ShutDown()
+				return
+			case <-sigChan:
+				queue.AddRateLimited(struct{}{})
+			}
+		}
+	}()
 
-				select {
-				case <-stopCh:
-					close(sigChan)
-					return
+	go func() {
+		for {
+			setupLog.Info("Launching watcher for ConfigMaps")
+			if err := configWatcher.Watch(sigChan, stopCh); err != nil {
+				switch err {
+				case watcher.ErrWatchChannelClosed:
+					// known issue: https://github.com/kubernetes/client-go/issues/334
+					setupLog.Info("watcher channel has closed, restart watcher")
 				default:
-					// sleep 2 seconds to prevent excessive log due to infinite restart
-					time.Sleep(2 * time.Second)
+					setupLog.Error(err, "unable to watch new ConfigMaps")
+					os.Exit(1)
 				}
 			}
-		}()
 
-		for {
 			select {
-			case <-eventsCh:
-				setupLog.Info("Triggering ConfigMap reconciliation")
-				updatedInjectionConfigs, err := configWatcher.GetInjectionConfigs()
-				if err != nil {
-					setupLog.Error(err, "unable to get ConfigMaps")
-					continue
-				}
-
-				setupLog.Info("Updating server with newly loaded configurations",
-					"original configs count", len(cfg.Injections), "updated configs count", len(updatedInjectionConfigs))
-				cfg.ReplaceInjectionConfigs(updatedInjectionConfigs)
-				setupLog.Info("Configuration replaced")
 			case <-stopCh:
-				break
+				close(sigChan)
+				return
+			default:
+				// sleep 2 seconds to prevent excessive log due to infinite restart
+				time.Sleep(2 * time.Second)
 			}
 		}
-
 	}()
+
+	return queue
+}
+
+func watchConfig(configWatcher *watcher.K8sConfigMapWatcher, cfg *config.Config, stopCh <-chan struct{}) {
+	queue := setupWatchQueue(stopCh, configWatcher)
+
+	for {
+		item, shutdown := queue.Get()
+		defer queue.Done(item)
```

Comment:
I think we should not use `defer` in a for "loop", it will add the defer logic for outer function `watchConfig`, not for each loop.

We could use a closure `func() {}()` for using defer in the for loop.

```go
for {
  if item, shutdown := queue.Get(); shutdown {
    break
  }
  func (){
    defer queue.Done(item)
    // other codes here
  }()
}
```

Reply:
Fixed